Added the WC Admin enabled db update notice.

peterfabian opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/25646 .

### How to test the changes in this Pull Request:

1. Create a package out of this branch.
2. Create a new JN site, install and activate WC, load sample products and create an order.
3. Upgrade WC using package from step 1.
4. There should be 3 new notices on the WC Admin pages (e.g. Orders, Coupons, etc), while the old notices appear on all the other pages (e.g. list of plugins). They should correspond 1:1, i.e. when the old "Update in progress" notice appears, the new one should also appear and vice versa. If the new notice is dismissed, old ones should no longer appear, and vice versa. 

DB update needed, notice 1 of 3:
<img width="1678" alt="Screenshot 2020-02-23 at 22 12 15" src="https://user-images.githubusercontent.com/2207451/75120167-b37c5680-5689-11ea-8982-6660b43d69b3.png">

DB update in progress, notice 2 of 3:
<img width="1678" alt="Screenshot 2020-02-23 at 22 12 27" src="https://user-images.githubusercontent.com/2207451/75120173-becf8200-5689-11ea-9015-9266b7687e49.png">

DB update done, notice 3 of 3:
<img width="1678" alt="Screenshot 2020-02-23 at 22 12 56" src="https://user-images.githubusercontent.com/2207451/75120176-c68f2680-5689-11ea-9eb6-da936dca54a4.png">


4. Test this with WP 5.3 and WP 5.2 to make sure old notices still work in WP 5.2 and new ones appear with WC Admin on WP 5.3.
5. After updating, try to change the db version back to see if the notice gets triggered again (e.g. run `UPDATE wp_options SET option_value='3.9.2' where option_name='woocommerce_db_version';`)

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Added new db update notice for WC Admin screens.

peterfabian committed:
Added the WC Admin enabled db update notice.

vedanshujain commented on includes/admin/notes/class-wc-notes-run-db-update.php:
```diff
@@ -0,0 +1,312 @@
+	private static function should_show_notice() {
+		if ( ! \WC_Install::needs_db_update() ) {
+			try {
+				$data_store = \WC_Data_Store::load( 'admin-note' );
```
note that this needs wc-admin tables to be installed. so we may not see notices immediately after activation if table generation script is still running or (maybe) if plugin is updated via ftp instead of wp-admin. have we tested this scenario?

peterfabian replied:
Good point, checking.

vedanshujain replied:
Looks like intial db install happens synchronously on first request after updation, so I guess we should be here. I tested out by replacing WC 3.9 without wc-admin with WC 4.0 via SFT and faced no issues. Resolving this conversation.

vedanshujain commented on includes/admin/notes/class-wc-notes-run-db-update.php:
```diff
@@ -0,0 +1,312 @@
+			return;
+		}
+
+		return current( $note_ids );
```
are we returning the last note_id? lets be more explicit here. WDYT?

peterfabian replied:
In either case, there should be max 1 of these notes in the db, so I'm returning the 'current' one. Maybe I should add some cleanup to make sure there are no more instances of this note in the db, then it should not matter?

vedanshujain replied:
If its usually 1 then I guess current implementation is fine. This is just a nitpick but I find `current` hard to read if don't see any iterations on the list nearby. 
Lets keep the current implementation for now.

vedanshujain commented on includes/class-wc-install.php:
```diff
@@ -248,6 +260,11 @@ public static function install_actions() {
 			check_admin_referer( 'wc_db_update', 'wc_db_update_nonce' );
 			self::update();
 			WC_Admin_Notices::add_notice( 'update' );
+			if ( WC()->is_wc_admin_active() ) {
```
may be also add a check for if wc_admin is installed as well?

peterfabian replied:
I might be missing something, but it seems to me that if it's not installed, then it's not active (the method checks whether a function exists, so it only returns true if installed && active)?

vedanshujain replied:
I think I meant we can check here if tables are also installed. Let me actually test this scenario and the one described in https://github.com/woocommerce/woocommerce/pull/25736#discussion_r383253956

vedanshujain replied:
Resolving based on testing.

vedanshujain reviewed (commented):
Tested out and working as expected. There are few edge case that I have not tested, but had some comments to wanted to post them soonish.

peterfabian commented on includes/admin/notes/class-wc-notes-run-db-update.php:
```diff
@@ -0,0 +1,312 @@
+	 * @return bool
+	 */
+	private static function should_show_notice() {
+		if ( ! \WC_Install::needs_db_update() ) {
```
Please see the function comment. 

If the db does not need an update, but the notice has *not* been actioned (i.e. after the db update, when 
store owner hasn't acknowledged the successful db update (notice 3 out of 3)), still show the notice.

vedanshujain replied:
ah my bad. resolving the conversation.

peterfabian committed:
Added more exclusions for the DB notice show_reminder method.

Also, moved `woocommerce_hide_update_notice` outside those exclusions in case
it would get triggered from cron or via AJAX.

peterfabian commented on includes/admin/notes/class-wc-notes-run-db-update.php:
```diff
@@ -0,0 +1,312 @@
+
+		add_action( 'admin_init', array( __CLASS__, 'show_reminder' ) );
+		// If the old notice gets dismissed, also hide this new one.
+		add_action( 'woocommerce_hide_update_notice', array( __CLASS__, 'set_notice_actioned' ) );
```
Agree, addressed in https://github.com/woocommerce/woocommerce/pull/25736/commits/f3dceae5dff179c4a370b6a5031acf65f4e0253a.

peterfabian committed:
Force the saving of notices only for the 'update' notice to prevent performance problems.

peterfabian committed:
Force the saving of notices for the 'install' notice to prevent performance problems for all notices, while also making it less prone to race conditions during the installation.

peterfabian commented on includes/admin/class-wc-admin-notices.php:
```diff
@@ -105,6 +105,9 @@ public static function reset_admin_notices() {
 		self::$notices = array_unique( array_merge( self::get_notices(), array( $name ) ) );
+
+		// Adding early save to prevent more race conditions with notices.
+		self::store_notices();
```
Agree, that might be a good idea. Should be solved by https://github.com/woocommerce/woocommerce/pull/25736/commits/992c6538dcc2d57f5c7ffbc140b31517f1d603ac and https://github.com/woocommerce/woocommerce/pull/25736/commits/63486a1cb06c9d183355a2d3162ca18b719578cf

peterfabian commented on includes/admin/notes/class-wc-notes-run-db-update.php:
```diff
@@ -0,0 +1,312 @@
+	 * Attach hooks.
+	 */
+	public function __construct() {
+		if ( defined( 'DOING_AJAX' ) && DOING_AJAX ) {
```
Added more conditions in f3dceae, checking the loading order.

peterfabian replied:
Tested this now by removing the `wp_wc_admin_notes` db table and the code just runs with no ill side effects, with $note_id == 0. The reminder will only show up once the db tables for WC Admin get created, which should happen during the first `admin_init` in `\Automattic\WooCommerce\Admin\Install::check_version` (or during subsequent `admin_init` in case the version of wc admin changes).

vedanshujain reviewed (approved):
LGTM!
